core/trashcan.js

```diff
@@ -467,35 +459,28 @@ Blockly.Trashcan.prototype.mouseOut_ = function() {
 
 /**
  * Handle a BLOCK_DELETE event. Adds deleted blocks oldXml to the content array.
- * @return {!Function} Function to call when a block is deleted.
+ * @param {!Blockly.Events.Abstract} event Workspace event.
  * @private
  */
-Blockly.Trashcan.prototype.onDelete_ = function() {
-  var trashcan = this;
-  return function(event) {
-    if (trashcan.workspace_.options.maxTrashcanContents <= 0) {
+Blockly.Trashcan.prototype.onDelete_ = function(event) {
+  if (this.workspace_.options.maxTrashcanContents <= 0) {
+    return;
+  }
+  if (event.type == Blockly.Events.BLOCK_DELETE &&
+      event.oldXml.tagName.toLowerCase() != 'shadow') {
+    var cleanedXML = this.cleanBlockXML_(event.oldXml);
+    if (this.contents_.indexOf(cleanedXML) != -1) {
       return;
     }
-    if (event.type == Blockly.Events.BLOCK_DELETE &&
-        event.oldXml.tagName.toLowerCase() != 'shadow') {
-      var cleanedXML = trashcan.cleanBlockXML_(event.oldXml);
-      if (trashcan.contents_.indexOf(cleanedXML) != -1) {
-        return;
-      }
-      trashcan.contents_.unshift(cleanedXML);
-      if (trashcan.contents_.length >
-          trashcan.workspace_.options.maxTrashcanContents) {
-        trashcan.contents_.splice(
```

Comment:
A single splice is more efficient than using a while loop and popping off multiple unwanted items.  But in Blockly's trashcan, n will always be 1.  So if/pop is simpler and while/pop adds a level of safety.